src/repo/fs/blocks.rs

```diff
@@ -100,6 +100,58 @@ impl FsBlockStore {
             }
         }
     }
+
+    /// Implementation moved here from inside the `impl BlockStore for FsBlockStore` which is
+    /// `#[async_trait]`, and somehow the two lifetimes generated by `#[async_trait]` just don't
+    /// work leading to confusing error when `ipfs` is used as a dependency in an outside of
+    /// workspace crate.
+    ///
+    /// See [gist] for the standalone working workaround, error and an expanded version of the
+    /// original code.
+    ///
+    /// [gist]: https://gist.github.com/koivunej/d6abccb4133839eeab8b36992f1a95fa
+    async fn list0(&self) -> Result<Vec<Cid>, Error> {
+        use futures::future::{ready, Either};
+        use futures::stream::{empty, TryStreamExt};
+        use tokio_stream::wrappers::ReadDirStream;
+
+        let span = tracing::trace_span!("listing blocks");
+
+        async move {
+            let stream = ReadDirStream::new(fs::read_dir(self.path.clone()).await?);
+
+            // FIXME: written as a stream to make the Vec be BoxStream<'static, Cid>
+            let vec = stream
+                .and_then(|d| async move {
+                    // map over the shard directories
+                    Ok(if d.file_type().await?.is_dir() {
+                        Either::Left(ReadDirStream::new(fs::read_dir(d.path()).await?))
+                    } else {
+                        Either::Right(empty())
+                    })
+                })
```

Comment:
this could be simplified to another `try_filter_map` which would allow less types, probably faster compilation.